controller/backup_backing_image_controller.go

```diff
@@ -0,0 +1,478 @@
+package controller
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/rancher/lasso/pkg/log"
+	"github.com/sirupsen/logrus"
+
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	clientset "k8s.io/client-go/kubernetes"
+	v1core "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/kubernetes/pkg/controller"
+
+	"github.com/longhorn/backupstore/backupbackingimage"
+
+	"github.com/longhorn/longhorn-manager/datastore"
+	"github.com/longhorn/longhorn-manager/engineapi"
+	"github.com/longhorn/longhorn-manager/types"
+	"github.com/longhorn/longhorn-manager/util"
+
+	longhorn "github.com/longhorn/longhorn-manager/k8s/pkg/apis/longhorn/v1beta2"
+)
+
+type BackupBackingImageController struct {
+	*baseController
+
+	// which namespace controller is running with
+	namespace string
+	// use as the OwnerID of the controller
+	controllerID string
+
+	kubeClient    clientset.Interface
+	eventRecorder record.EventRecorder
+
+	monitors    map[string]*engineapi.BackupBackingImageMonitor
+	monitorLock sync.RWMutex
+
+	ds *datastore.DataStore
+
+	cacheSyncs []cache.InformerSynced
+
+	proxyConnCounter util.Counter
+}
+
+func NewBackupBackingImageController(
+	logger logrus.FieldLogger,
+	ds *datastore.DataStore,
+	scheme *runtime.Scheme,
+	kubeClient clientset.Interface,
+	controllerID string,
+	namespace string,
+	proxyConnCounter util.Counter,
+) *BackupBackingImageController {
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(logrus.Infof)
+	// TODO: remove the wrapper when every clients have moved to use the clientset.
+	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{
+		Interface: v1core.New(kubeClient.CoreV1().RESTClient()).Events(""),
+	})
+
+	bc := &BackupBackingImageController{
+		baseController: newBaseController("longhorn-backup-backing-image", logger),
+
+		namespace:    namespace,
+		controllerID: controllerID,
+
+		monitors:    map[string]*engineapi.BackupBackingImageMonitor{},
+		monitorLock: sync.RWMutex{},
+
+		ds: ds,
+
+		kubeClient:    kubeClient,
+		eventRecorder: eventBroadcaster.NewRecorder(scheme, v1.EventSource{Component: "longhorn-backup-backing-image-controller"}),
+
+		proxyConnCounter: proxyConnCounter,
+	}
+
+	ds.BackupBackingImageInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{ //nolint:errcheck
+		AddFunc:    bc.enqueueBackupBackingImage,
+		UpdateFunc: func(old, cur interface{}) { bc.enqueueBackupBackingImage(cur) },
+		DeleteFunc: bc.enqueueBackupBackingImage,
+	})
```

Comment:
~Why using periodic sync? Use AddEventHandlerWithResyncPeriod with 0 resync instead.~

Ok, you are relying on the periodic sync to update the status from the monitor.